crypto.go

```diff
@@ -34,37 +43,68 @@ func Equals(sk, other []byte) bool {
 
 // Verify checks the given signature and returns true if it is valid.
 func Verify(pk, msg, signature []byte) bool {
+	pub, err := secpEc.NewPublicKey(pk)
+	if err != nil {
+		return false
+	}
+
+	ecsdaOpts := &secpEc.ECDSAOptions{Encoding: secpEc.EncodingCompact, RejectMalleable: true, Hash: crypto.SHA256}
+
+	// Check if the signature is in [R | S | V] style.
+	// if so, we change the Encoding to "EncodingCompactRecoverable"
 	if len(signature) == 65 {
-		// Drop the V (1byte) in [R | S | V] style signatures.
+		// Normally you would drop the V (1byte) in [R | S | V] style signatures.
 		// The V (1byte) is the recovery bit and is not apart of the signature verification.
-		return secp256k1.VerifySignature(pk[:], msg, signature[:len(signature)-1])
+		// EncodingCompactRecoverable expects the V to be included in the signature
+		ecsdaOpts.Encoding = secpEc.EncodingCompactRecoverable
 	}
 
-	return secp256k1.VerifySignature(pk[:], msg, signature)
+	return pub.Verify(msg, signature, ecsdaOpts)
```

Comment:
I would keep doing this the old way. I.e., drop the recovery bit then validate as `EncodingCompact`. I think it's technically slower than the recovery method... but I'd prefer to keep the consistency with the old logic.

Reply:
@Stebalien I've implemented this fix in [this commit](https://github.com/filecoin-project/go-crypto/pull/5/commits/2f64eabd5e6673d007e10ddac316c962f8125940), but I'm still questioning if we're introducing any signature malleability risk by doing it this way. I doubt we would introduce this risk when `secpEc.ECSDAOptions.RejectMalleable == true`, but I'm not sure what effect stripping the recoveryID on a signature that originally has a recoveryID would create. 

The `RejectMalleable` option, according to the lib's code comments:

```
// RejectMalleable will cause the verification process to
// reject signatures where `s > n / 2`.
```

But will this protect against all the same signature malleability attacks that could come from ignoring the recoveryID all together?

Reply:
I think you're right. What tripped me up is:

1. On-chain, we _always_ use `EcRecover` (which uses the recovery byte).
2. As far as I can tell, we _never_ call `Verify` _anywhere_.

So... honestly, I'd drop support for 64-byte long signatures and _always_ `EcRecover` inside `Verify`. That way we only have one signature format.

Reply:
(and thanks for pushing back on this, I was definitely wrong)